src/storage/lock_manager/lock_waiting_queue.rs

```diff
@@ -0,0 +1,950 @@
+    /// `KeyLockWaitState` is:
+    ///
+    /// ```text
+    /// legacy_wake_up_cnt: 0, queue: [(20, 0), (30, 0), (40, 0)]
+    /// ```
+    ///
+    /// Then the lock on the key is released. We pops the first entry in the
+    /// queue to wake it up, and then schedule a call to
+    /// [`LockWaitQueues::delayed_notify_all`] after delaying for
+    /// `wake_up_delay_duration`. The current state becomes:
+    ///
+    /// ```text
+    /// legacy_wake_up_cnt: 1, queue: [(30, 0), (40, 0)]
+    /// ````
+    ///
+    /// Here, if some other request arrives, one of them may successfully
+    /// acquire the lock and others are pushed to the queue. the state
+    /// becomes:
+    ///
+    /// ```text
+    /// legacy_wake_up_cnt: 1, queue: [(30, 0), (40, 0), (50, 1), (60, 1)]
+    /// ```
+    ///
+    /// Then `wake_up_delay_duration` is elapsed since the previous waking up.
+    /// Here, we expect that the lock wait entries 30 and 40 can be woken
+    /// up, since they exists when the previous waking up occurs. But we
+    /// don't want to wake up later-arrived entries (50 and 60) since it
+    /// introduces useless pessimistic retries to transaction 50 and 60 when
+    /// they don't need to. The solution is, only wake up the entries that
+    /// has `entry.current_legacy_wake_up_cnt <
+    /// key_lock_wait_state.legacy_wake_up_cnt`. Therefore, we only wakes up
+    /// entries 30 and 40 who has `current_legacy_wake_up_cnt < 1`, while 50 and
+    /// 60 will be left untouched.
+    ///
+    /// When waking up resumable requests, the mechanism above won't take
+    /// effect. If a legacy request is woken up and triggered the mechanism,
+    /// and there is a resumable request in the queue, `delayed_notify_all`
+    /// will stop at the first resumable request it meets, pop it out, and
+    /// return it from a [`DelayedNotifyAllFuture`]. See
+    /// [`LockWaitQueues::pop_for_waking_up`].
+    legacy_wake_up_cnt: usize,
+    queue: BinaryHeap<Box<LockWaitEntry>>,
+
+    /// The start_ts of the most recent waking up event.
+    last_conflict_start_ts: TimeStamp,
+    /// The commit_ts of the most recent waking up event.
+    last_conflict_commit_ts: TimeStamp,
+
+    /// `(id, start_time, delay_duration)`
+    delayed_notify_all_state: Option<(u64, Instant, Arc<AtomicU64>)>,
+}
+
+impl KeyLockWaitState {
+    fn new() -> Self {
+        Self {
+            current_lock: kvrpcpb::LockInfo::default(),
+            legacy_wake_up_cnt: 0,
+            queue: BinaryHeap::new(),
+            last_conflict_start_ts: TimeStamp::zero(),
+            last_conflict_commit_ts: TimeStamp::zero(),
+            delayed_notify_all_state: None,
+        }
+    }
+}
+
+pub type DelayedNotifyAllFuture = Pin<Box<dyn Future<Output = Option<Box<LockWaitEntry>>> + Send>>;
+
+pub struct LockWaitQueueInner {
+    queue_map: dashmap::DashMap<Key, KeyLockWaitState>,
+    id_allocated: AtomicU64,
```

Comment:
It's used to allocate ID for `KeyLockWaitState`. The ID is used in `delayed_notify_all` to prevent the following case:

1. Waking up occurs on key A, and there are remaining lock-waiting requests so schedules a call to `delayed_notify_all`
2. Waking up continues occurring on key A and the queue is cleared, so that the key is removed from the queue map.
3. Lock wait occurs on key A, and a new KeyLockWaitState (with initial legacy_wake_up_cnt = 0) is added to the map.
4. `delayed_notify_all` scheduled at step 1 is called, but nothing in the current queue should be notified.

With this ID, the previously-removed `KeyLockWaitState` and the later inserted one can be distinguished. So in this case the `delayed_notify_all` invocation at step 4 does nothing.

Reply:
Sorry, I explained it wrong. The id is not attached to the `KeyLockWaitState`, but will be allocated when scheduling a new `delayed_notify_all`, and is stored in the `delayed_notify_all_state` field in the `KeyLockWaitState`. It may change when a `delayed_notify_all` is done and a new one is scheduled. But it performs the same functionality as I described above.